src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/FormattingPassBase.cs

```diff
@@ -224,81 +317,128 @@ protected static SourceText CleanupDocument(FormattingContext context, Range ran
-                if (whitespaceLength == null)
-                {
-                    // There was no content here. Skip.
-                    continue;
-                }
+                return;
+            }
 
-                var spanToReplace = new TextSpan(mappingSpan.Start, whitespaceLength.Value);
-                if (!context.TryGetIndentationLevel(spanToReplace.End, out var contentIndentLevel))
-                {
-                    // Can't find the correct indentation for this content. Leave it alone.
-                    continue;
-                }
+            if (!ShouldFormat(context, sourceMappingRange.End, allowImplicitStatements: false))
+            {
+                // We don't want to run cleanup on this range.
+                return;
+            }
 
-                // At this point, `contentIndentLevel` should contain the correct indentation level for `}` in the above example.
-                var replacement = context.NewLineString + context.GetIndentationLevelString(contentIndentLevel);
+            var contentStartOffset = text.Lines[mappingEndLineIndex].GetFirstNonWhitespaceOffset(sourceMappingRange.End.Character);
+            if (contentStartOffset == null)
+            {
+                // There is no content after the end of this source mapping. No need to clean up.
+                return;
+            }
 
-                // After the below change the above example should look like,
-                // @{
-                //     if (true)
-                //     {     
-                //         <div></div>
-                //     }
-                // }
-                var change = new TextChange(spanToReplace, replacement);
-                changes.Add(change);
+            var spanToReplace = new TextSpan(sourceMappingSpan.End, 0);
+            if (!context.TryGetIndentationLevel(spanToReplace.End, out var contentIndentLevel))
+            {
+                // Can't find the correct indentation for this content. Leave it alone.
+                return;
             }
 
-            var changedText = text.WithChanges(changes);
-            return changedText;
+            // At this point, `contentIndentLevel` should contain the correct indentation level for `}` in the above example.
+            var replacement = context.NewLineString + context.GetIndentationLevelString(contentIndentLevel);
+
+            // After the below change the above example should look like,
+            // @{
+            //     if (true)
+            //     {
+            //         <div></div>
+            //     }
+            // }
+            var change = new TextChange(spanToReplace, replacement);
+            changes.Add(change);
         }
 
-        private static bool ShouldCleanup(FormattingContext context, Position position)
+        protected bool ShouldFormat(FormattingContext context, Position position, bool allowImplicitStatements)
```

Comment:
> implicit statement or implicit expression?

Statements like `@if (true){...}`. 

> is the reason why the flag is even here is because of the differences between OnTypeFormatting and regular Formatting

That too. But this is mainly the difference between whether to format using the C# formatter and whether to cleanup that source mapping. They both use `ShouldFormat` to decide whether they should operate on it. The job of CleanupDocument is to introduce a newline at the start and end of each source mapping. But we can do that for implicit statements because then it would be wrong like below,
```
@
if(true){...}
```